debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ChangeStreamPipelineFactory.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.bson.conversions.Bson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.client.model.Aggregates;
+import com.mongodb.client.model.Field;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.changestream.OperationType;
+
+import io.debezium.connector.mongodb.Filters.FilterConfig;
+import io.debezium.data.Envelope;
+
+/**
+ * A factory to produce a MongoDB change stream pipeline expression.
+ */
+class ChangeStreamPipelineFactory {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChangeStreamPipelineFactory.class);
+
+    private final ReplicaSetOffsetContext rsOffsetContext;
+    private final MongoDbConnectorConfig connectorConfig;
+    private final FilterConfig filterConfig;
+
+    ChangeStreamPipelineFactory(ReplicaSetOffsetContext rsOffsetContext, MongoDbConnectorConfig connectorConfig, FilterConfig filterConfig) {
+        this.rsOffsetContext = rsOffsetContext;
+        this.connectorConfig = connectorConfig;
+        this.filterConfig = filterConfig;
+    }
+
+    List<Bson> create() {
+        // Resolve the leaf filters
+        var filters = Stream
+                .of(
+                        createCollectionFilter(filterConfig),
+                        createOperationTypeFilter(connectorConfig),
+                        createClusterTimeFilter(rsOffsetContext))
+                .flatMap(Optional::stream)
+                .collect(toList());
+
+        // Combine
+        var andFilter = Filters.and(filters);
+        var matchFilter = Aggregates.match(andFilter);
+
+        // Pipeline
+        // Note that change streams cannot use indexes:
+        // - https://www.mongodb.com/docs/manual/administration/change-streams-production-recommendations/#indexes-and-performance
+        // Note that `$addFields` must be used over `$set`/ `$unset` to support MongoDB 4.0 which doesn't support these operators:
+        // - https://www.mongodb.com/docs/manual/changeStreams/#modify-change-stream-output
+        var pipeline = List.of(
+                addFields("namespace", concat("$ns.db", ".", "$ns.coll")), // TODO : This needs to come after the change type filtering since some events don't have these fields!
```

Comment:
Per the [docs](https://www.mongodb.com/docs/manual/reference/operator/aggregation/concat/#mongodb-expression-exp.-concat):

<img width="789" alt="image" src="https://user-images.githubusercontent.com/2717578/220961183-25c6dcf5-e130-464b-84d9-8bc2e5f03faf.png">

This is also true of [4.0](https://www.mongodb.com/docs/v4.0/reference/operator/aggregation/concat/), so we should be good 👍 .